control-plane/connect-inject/peering_dialer_controller.go

```
@@ -0,0 +1,268 @@
+			controllerutil.AddFinalizer(dialer, FinalizerName)
+			if err := r.Update(ctx, dialer); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if containsString(dialer.Finalizers, FinalizerName) {
+			r.Log.Info("PeeringDialer was deleted, deleting from Consul", "name", req.Name, "ns", req.Namespace)
+			if err := r.deletePeering(ctx, req.Name); err != nil {
+				return ctrl.Result{}, err
+			}
+			controllerutil.RemoveFinalizer(dialer, FinalizerName)
+			err := r.Update(ctx, dialer)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Get the status secret and the spec secret.
+	// Cases need to handle secretRefSet, statusSecret, secretSet, specSecret.
+	// no secretSet --> error bc spec needs to be set.
+	// no specSecret --> error bc waiting for spec secret to exist.
+	// no secretRefSet, yes secretSet, no statusSecret, yes specSecret --> initiate peering.
+	// yes secretRefSet, yes secretSet, no statusSecret, yes specSecret --> initiate peering.
+	// yes secretRefSet, yes secretSet, yes statusSecret, yes specSecret --> compare contents, if
+	// different initiate peering.
+
+	// TODO(peering): remove this once CRD validation exists.
+	secretSet := false
+	if dialer.Secret() != nil {
+		secretSet = true
+	}
+	if !secretSet {
+		err = errors.New("PeeringDialer spec.peer.secret was not set")
+		r.updateStatusError(ctx, dialer, err)
+		return ctrl.Result{}, err
+	}
+
+	// specSecret will be nil if the secret specified by the spec doesn't exist.
+	var specSecret *corev1.Secret
+	specSecret, err = r.getSecret(ctx, dialer.Secret().Name, dialer.Namespace)
+	if err != nil {
+		r.updateStatusError(ctx, dialer, err)
+		return ctrl.Result{}, err
+	}
+
+	// If specSecret doesn't exist, error because we can only initiate peering if we have a token to initiate with.
+	if specSecret == nil {
+		err = errors.New("PeeringDialer spec.peer.secret does not exist")
+		r.updateStatusError(ctx, dialer, err)
+		return ctrl.Result{}, err
+	}
```

should `getSecret` return an error for this case?

---

🤔 In this case though we're getting the secret from the spec not status. And the call above on line 88 gets the secret. Then here we're checking if secret is nil and returning an error, so I'm wondering why not combine these two calls into one and just make `getSecret` return an error when the secret doesn't exist instead of returning an error here.

Maybe I'm misunderstanding something.

---

Aah! sorry. I should have been specific. We use the method to get the status secret (L110) too. And if it is the first reconcile, it is ok if the secret doesn't exist, so it would be ok for getSecret to not return an error if the secret does not exist.